### PR TITLE
RavenDB-19557 added equality comparer for distinct operations to handle case when enumerable has mixed LSV and string values

### PR DESCRIPTION
**Pull request by ppekrol**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19557

### Type of change

- Bug fix

### How risky is the change?

- Moderate 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/Static/DynamicArray.cs [hunk 1]
```diff
@@ -13,6 +13,8 @@ public class DynamicArray : DynamicObject, IOrderedEnumerable<object>
     {
         private readonly IEnumerable<object> _inner;
 
+        private Lazy<JsonOperationContext> _context = new(() => CurrentIndexingScope.Current?.IndexContext);
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Static/DynamicArray.cs]: Why is this lazy?

Comment (ppekrol) [reply src/Raven.Server/Documents/Indexes/Static/DynamicArray.cs]: No readon, for me the option without Lazy is more dirty, pushed

Comment (ppekrol) [reply src/Raven.Server/Documents/Indexes/Static/DynamicArray.cs]: ...or less readable

File: src/Raven.Server/Documents/Indexes/Static/DynamicArray.cs [hunk 2]
```diff
@@ -585,7 +585,7 @@ public IEnumerable<object> Where(Func<object, int, bool> func)
 
         public IEnumerable<object> Distinct()
         {
-            return new DynamicArray(Enumerable.Distinct(this));
+            return new DynamicArray(Enumerable.Distinct(this, new LazyStringEqualityComparer(CurrentIndexingScope.Current?.IndexContext)));
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Static/DynamicArray.cs]: Can we use a single static instance here?

File: src/Raven.Server/Documents/Indexes/Static/DynamicArray.cs [hunk 3]
```diff
@@ -840,5 +840,62 @@ IEnumerator IEnumerable.GetEnumerator()
                 return GetEnumerator();
             }
         }
+
+        private class LazyStringEqualityComparer : IEqualityComparer<object>
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/Static/DynamicArray.cs]: I think the name might be a bit misleading. It's not the comparer which is supposed to be used to compare LazyStringValues only. It's a comparer of any objects which has special behavior of dealing with string and LSV. Also currently it's used for distinct purposes only. Maybe `LazyStringAwareDistinctComparer`?

**Review by arekpalinski — APPROVED** [review]

**Pull request merged**